src/librbd/cache/ParentCacheObjectDispatch.cc

```diff
@@ -77,29 +70,13 @@ bool ParentCacheObjectDispatch<I>::read(
   auto cct = m_image_ctx->cct;
   ldout(cct, 20) << "object_no=" << object_no << " " << object_off << "~"
                  << object_len << dendl;
-  ceph_assert(m_initialized);
   string oid = data_object_name(m_image_ctx, object_no);
 
   /* if RO daemon still don't startup, or RO daemon crash,
    * or session occur any error, try to re-connect daemon.*/
+  std::unique_lock locker{m_lock};
   if (!m_cache_client->is_session_work()) {
```

Comment:
@dillaman I am not sure we can get rid of `connecting` variable entirely. For me it just seemed useless to have it atomic after we had introduced the lock.

As I see in the cache client code, the `session_work` is set to true only in `register_client`, i.e. far after we drop the lock. So it looks like now two requests may call `create_cache_session` concurrently. Don't you see any problem with this?

Reply:
Re-added and tweaked some more.